packages/modules/imodel-browser/src/containers/iModelGrid/useIModelData.ts

```diff
@@ -32,148 +32,227 @@ export const useIModelData = ({
+  const [abortController, setAbortController] = React.useState<
+    AbortController | undefined
+  >(undefined);
+  const updateAbortController = React.useCallback(
+    (val?: AbortController) => {
+      if (abortController) {
+        abortController.abort();
+      }
+      return val;
+    },
+    [abortController]
+  );
+
   const sortType =
     sortOptions && ["name", "createdDateTime"].includes(sortOptions.sortType)
       ? sortOptions.sortType
-      : undefined; //Only available sort by API at the moment.
+      : undefined; //Only available sort-by API at the moment.
+  const [previousSortOptions, setPreviousSortOptions] = React.useState<
+    IModelSortOptions | undefined
+  >(sortOptions && { ...sortOptions });
   const sortDescending = sortOptions?.descending;
-  const [iModels, setIModels] = React.useState<IModelFull[]>([]);
   const sortedIModels = useIModelSort(iModels, sortOptions);
-  const [status, setStatus] = React.useState<DataStatus>();
-  const [page, setPage] = React.useState(0);
-  const [morePages, setMorePages] = React.useState(true);
-  const fetchMore = React.useCallback(() => {
-    viewMode === "cells" && setStatus(DataStatus.Fetching);
-    status !== DataStatus.Fetching && setPage((page) => page + 1);
-  }, [status, viewMode]);
+  const sortChanged =
+    sortOptions?.descending !== previousSortOptions?.descending ||
+    sortOptions?.sortType !== previousSortOptions?.sortType;
+  if (sortChanged) {
+    setPreviousSortOptions(sortOptions);
+  }
 
-  React.useEffect(() => {
-    // If sort changes but we already have all the data,
-    // let client side sorting do its job, otherwise, refetch from scratch.
-    if (morePages) {
-      setStatus(DataStatus.Fetching);
-      setIModels([]);
-      setPage(0);
-      setMorePages(true);
+  const fetchMore = React.useCallback(() => {
+    if (status === DataStatus.Fetching || !morePagesAvailable) {
+      return;
     }
-  }, [sortType, sortDescending, morePages]);
+    setPage((prev) => prev + 1);
+    setNeedsUpdate(true);
+  }, [status, morePagesAvailable]);
+
   React.useEffect(() => {
-    // If any of the dependencies change, always restart the fetch from scratch.
+    // start from scratch when any external state changes
     setStatus(DataStatus.Fetching);
     setIModels([]);
     setPage(0);
-    setMorePages(true);
+    setMorePagesAvailable(true);
+    setNeedsUpdate(true);
   }, [
-    accessToken,
     iTwinId,
+    accessToken,
+    sortOptions?.descending,
+    sortOptions?.sortType,
     apiOverrides?.data,
     apiOverrides?.serverEnvironmentPrefix,
     searchText,
     maxCount,
   ]);
 
+  // Main function
   React.useEffect(() => {
-    if (!morePages) {
+    if (!needsUpdate) {
       return;
     }
+
+    setNeedsUpdate(false);
+    setAbortController(updateAbortController);
+
+    // if data is provided, use it and skip fetching
     if (apiOverrides?.data) {
-      setIModels(apiOverrides.data);
+      setIModels(apiOverrides?.data);
```

Comment:
nit: we already checked that apiOverrides.data isn't undefined so we don't need the optional chaining operator here.